Throw an Exception instead of an error on Cache::write() fail.

fr3nch13 opened the pull request:
### Summary: 
Throw a CacheWriteException when a Cache::write() fails, instead of triggering an error. This allows the developer more control over how to handle a Cache::write() issue.

### Details:
In my specific case, I use memcached as the primary cache engine. With a memcache server running internally/locally.
When I configure my cache engines, I'm setting the 'server' to '127.0.0.1/localhost'. In the development environments (locally, and on my dev server), and test them, it works fine.
However, when I run it through the CI process using a docker image, it fails as the 'server' should be set to 'memcached'.
I've tried to use the 'fallback' setting in the cache config to fall back to the 'default' config, which is using the FileEngine, but it triggers this error, instead of falling back to that config.
I'd like to be able to catch when a write fails, without needing to use the set_error_handler() function, so that I can account for it.
Situations like above, or if the memcached/redis server goes offline, or some other situation that would cause the cache to fail writing.

fr3nch13 committed:
Throw an Exception instead of a error on write fail.

Throw a CacheWriteException when a Cache::write() fails, instead of triggering an error.
This allows the developer more control over how to handle a Cache::write() issue.

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Update ci.yml

Moving the cs-stan job, and requiring the testsuites to have cs-stan be successful first.
This should catch cs issues before using other resources.

markstory commented on .github/workflows/ci.yml:
```diff
@@ -16,7 +16,55 @@ permissions:
+      run: vendor/bin/phpstan.phar analyse -c tests/phpstan.neon --error-format=github
+
   testsuite:
+    needs: cs-stan
```
Why did this need to change?

fr3nch13 replied:
I noticed when I submitted this pull request that all of the jobs ran in parallel. So it spun up the different environments for testing along with the job for CS and Stan checks. I made this change so that the CS and Stan checks would run, and only kick off the test suites is the Checks passed first. This is to keep the test suites from running if the only thing that needs to be fixed is something like the Use [classname]; list is out of order or something.
It basically just saves resources, especially if you pay for said resources to run. It saves those cycles.
My CI runs like:
phpcs -> phpstan -> phpunit

ADmad replied:
Please revert this revert this, we don't want to block the running of test suite in case of CS/stan failures. Getting info about failed tests earlier is more important for us.

markstory commented on src/Cache/CacheWriteException.php:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.7.0
```
```suggestion
 * @since         4.4.11
```

ADmad replied:
I don't think change should be done in a patch release. It should target 4.5.

fr3nch13 replied:
I would agree, as much as I'd like it sooner than later. :-p

othercorey commented:
Should we put all exceptions in an exception namespace?

fr3nch13 commented:
Oh ok. Do you need me to do it, or can you?On Jan 5, 2023, at 7:51 PM, ADmad ***@***.***> wrote:﻿
@ADmad commented on this pull request.



In .github/workflows/ci.yml:
>    testsuite:
+    needs: cs-stan

Please revert this revert this, we don't want to block the running of test suite in case of CS/stan failures. Getting info about failed tests earlier is more important for us.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

ADmad commented:
> Oh ok. Do you need me to do it, or can you?

You can just drop the commit in which ci.yml  is modified.

fr3nch13 committed:
Revert "Update ci.yml"

This reverts commit 9827f8b7a5a7db49f279e6e4887cd752209c489c.

othercorey commented:
@fr3nch13 Can you move the exception into an `Exception` namespace?

fr3nch13 commented:
@othercorey Where would you want to put it?
I put it in the Cache namespace as it was specific to that one possible problem, and the InvalidArguement one was already there.

othercorey commented:
> @othercorey Where would you want to put it? I put it in the Cache namespace as it was specific to that one possible problem, and the InvalidArguement one was already there.

`Cache\Exception`

We probably should move InvalidArgument into that namespace as well @markstory

fr3nch13 committed:
Moving Cache Exceptions to their own namespace.

fr3nch13 commented:
As per @othercorey 's comment, I moved the Cache Exceptions to their own namespace. I also noticed that php's core InvalidArgumentException was being used where the Cake\Cache\Exception\InvalidArgumentException should possibly be used. 
See: Cake\Cache\Engine\MemcachedEngine.php

fr3nch13 committed:
Oops. Fixed CS errors.

fr3nch13 commented on src/Cache/Engine/MemcachedEngine.php:
```diff
@@ -98,7 +98,7 @@ class MemcachedEngine extends CacheEngine
      *
      * @param array<string, mixed> $config array of setting for the engine
      * @return bool True if the engine has been successfully initialized, false if not
-     * @throws \InvalidArgumentException When you try use authentication without
```
Changed from using the Php exception to the cake one.

fr3nch13 commented on src/Cache/Engine/MemcachedEngine.php:
```diff
@@ -199,7 +199,7 @@ public function init(array $config = []): bool
      * Settings the memcached instance
      *
      * @return void
-     * @throws \InvalidArgumentException When the Memcached extension is not built
```
Changed from using the Php exception to the cake one.

markstory commented on src/Cache/Exception/InvalidArgumentException.php:
```diff
@@ -14,7 +14,7 @@
  * @license       https://opensource.org/licenses/mit-license.php MIT License
  */
-namespace Cake\Cache;
+namespace Cake\Cache\Exception;
```
We'll need to add a class_alias for the old class. Userland code could have `catch` blocks that reference the old class name and we can't break those references.

ADmad commented on src/Cache/Exception/CacheWriteException.php:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.7.0
```
Lies

fr3nch13 commented:
Lol, still getting used to updating the metadata. -------------Brian French703-507-5383On Jan 14, 2023, at 1:02 AM, ADmad ***@***.***> wrote:﻿
@ADmad commented on this pull request.



In src/Cache/Exception/CacheWriteException.php:
> @@ -0,0 +1,26 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.7.0

Lies

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

ADmad commented:
Needs to target 4.next

markstory committed:
Update version number to 4.4.11

markstory committed:
Add `class` template variable back.

Fixes #17008

markstory committed:
Apply suggestions from code review

Co-authored-by: Mark Scherer <dereuromark@users.noreply.github.com>

fr3nch13 committed:
Adding class alias as per Mark's recommendation

fr3nch13 committed:
Update InvalidArgumentException.php

fr3nch13 committed:
Merge branch 'cakephp:4.x' into 4.x

fr3nch13 committed:
Adding test for class alias.

fr3nch13 committed:
Update CacheTest.php

Try catching the aliased exception as would happen in apps.

fr3nch13 committed:
Try how class_alias is being used elsewhere here.

fr3nch13 committed:
Update InvalidArgumentException.php

fr3nch13 committed:
Update InvalidArgumentException.php

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Update CacheTest.php

Maybe phpcs will accept this? I really don't want to add to the neon file for this.

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Meta update.

fr3nch13 commented:
Maybe change it to the 4.4.12 milestone?

dereuromark committed:
Fix up null issue on PHP8.2+

fr3nch13 committed:
string check

Check to make sure the $template is specifically a string.

markstory commented on src/Cache/Exception/InvalidArgumentException.php:
```diff
@@ -25,3 +25,10 @@
 }
+// phpcs:disable
+// @deprecated Backwards compatibility alias. Will be removed in 5.0
+class_alias(
+    'Cake\Cache\Exception\InvalidArgumentException',
+    'Cake\Cache\InvalidArgumentException'
+);
```
Thanks for adding this. We'll also need a stub class in the old file location that contains a `class_exists('Cake\Cache\Exception\InvalidArgumentException')` in it. This will force the new class & the related alias to be loaded.

markstory committed:
Merge pull request #17012 from cakephp/4.x-null-issue

Fix up null issue on PHP8.2+

markstory committed:
Merge pull request #17013 from 302dev/issue-17011

Issue #17011

ravage84 committed:
Cast field name to string to prevent type hint issues with integer field names (#17006)

* Cast field name to string to prevent type hint issues with integer field names

* Cast integer field name to string to prevent type hint issues

* Add integers as possible types of field names in doc blocks

* Cover string casting of integer field names in Validator  with tests

* Do not cast int keys to string on field() & add()

* Fix syntax error

markstory committed:
Merge pull request #17009 from cakephp/fix-16980

Add `class` template variable back.

fr3nch13 committed:
Merge branch 'cakephp:4.x' into 4.x

fr3nch13 committed:
Adding stub file.

fr3nch13 committed:
Merge branch '4.next' into 4.x

fr3nch13 commented:
Closing this pull request as I've created another with the branches lining up correctly. See: #17034
